src/Compilers/Core/Portable/EncodedStringText.cs

```diff
@@ -170,30 +170,32 @@ private static byte[] TryGetByteArrayFromStream(Stream data)
         }
 
         /// <summary>
-        /// If the MemoryStream was created with publiclyVisible=true, then we can access its buffer
+        /// If the MemoryStream was created with publiclyVisible=true, then we can try to access its buffer
         /// directly and save allocations in StreamReader. The input MemoryStream is not closed on exit.
         /// </summary>
         /// <returns>True if a byte array could be created.</returns>
         private static bool TryGetByteArrayFromMemoryStream(MemoryStream data, out byte[] buffer)
         {
             Debug.Assert(data.Position == 0);
 
-            try
+            if (PortableShim.MemoryStream.TryGetBuffer != null)
             {
-                if (PortableShim.MemoryStream.GetBuffer != null)
+                ArraySegment<byte> bufferArraySegment;
+
+                if (PortableShim.MemoryStream.TryGetBuffer(data, out bufferArraySegment))
                 {
-                    buffer = (byte[])PortableShim.MemoryStream.GetBuffer.Invoke(data, null);
-                    return true;
+                    // since we can only return the whole array, make sure it's all used by the MemoryStream
+                    if (bufferArraySegment.Offset == 0 &&
+                        bufferArraySegment.Count == bufferArraySegment.Array.Length)
+                    {
```

Comment:
It should only be necessary to defend against offset != 0. The caller uses the stream length (which will be equal to the segment length) to decide how much of the buffer to use. It is very common for a memory stream to be backed by a buffer longer than its length. Not allowing that will regress the optimization.